File: octopipe/pkg/deployment/deployment.go
/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package deployment

import (
	"context"
	"octopipe/pkg/customerror"
	"os"
	"strconv"
	"time"

	"github.com/argoproj/gitops-engine/pkg/health"
	"github.com/argoproj/gitops-engine/pkg/utils/kube"
	"k8s.io/client-go/rest"
	"k8s.io/kubectl/pkg/cmd/util"

	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
)

const (
	DeployAction   = "DEPLOY"
	UndeployAction = "UNDEPLOY"
)

type UseCases interface {
	Do() error
	NewWatcher() error
	Deploy() error
	Undeploy() error
}

type Deployment struct {
	action    string
	update    bool
	namespace string
	manifest  map[string]interface{}
	config    *rest.Config
	kubectl   kube.Kubectl
}

func (main *DeploymentMain) NewDeployment(
	action string,
	update bool,
	namespace string,
	manifest map[string]interface{},
	config *rest.Config,
	kubectl kube.Kubectl,
) UseCases {
	return &Deployment{
		action:    action,
		update:    update,
		namespace: namespace,
		manifest:  manifest,
		config:    config,
		kubectl:   kubectl,
	}
}

func (deployment *Deployment) Do() error {
	switch deployment.action {
	case DeployAction:
		return deployment.watchDeploy()
	case UndeployAction:
		return deployment.Undeploy()
	default:
		return customerror.New("Deploy action failed", "Not recognize deploy action", nil, "deployment.Do")
	}
}

func getTimeoutDuration() time.Duration {
	defaultValue := time.Duration(300)
	envStr := os.Getenv("TIMEOUT_RESOURCE_VERIFICATION")
	if envStr == "" {
		return defaultValue
	}

	value, err := strconv.Atoi(envStr)
	if err != nil {
		return defaultValue
	}

	return time.Duration(value)
}

func (deployment *Deployment) NewWatcher() error {
	manifest := deployment.getUnstructuredManifest()
	ticker := time.NewTicker(3 * time.Second)
	timeout := time.After(getTimeoutDuration() * time.Second)
	for {
		select {
		case <-timeout:
			ticker.Stop()
			return customerror.New(
				"Resource watch failed",
				"Timeout",
				map[string]string{
					"resourceName":     manifest.GetName(),
					"groupVersionKind": manifest.GroupVersionKind().String(),
				},
				"deployment.newWatcher.timeout",
			)
		case <-ticker.C:

			gvk := manifest.GroupVersionKind()

			resource, err := deployment.kubectl.GetResource(context.TODO(), deployment.config, gvk, manifest.GetName(), deployment.namespace)
			if err != nil {
				return customerror.New(
					"Resource watch failed",
					err.Error(),
					map[string]string{
						"resourceName":     manifest.GetName(),
						"groupVersionKind": manifest.GroupVersionKind().String(),
					},
					"deployment.newWatcher.GetResource",
				)
			}

			if resource != nil {
				if !isResourController(resource) {
					ticker.Stop()
					return nil
				}

				healthStatus, err := health.GetResourceHealth(resource, nil)
				if err != nil {
					ticker.Stop()
					return customerror.New(
						"Resource watch failed",
						err.Error(),
						map[string]string{
							"resourceName":     manifest.GetName(),
							"groupVersionKind": manifest.GroupVersionKind().String(),
						},
						"deployment.newWatcher.GetResourceHealth",
					)
				}

				if healthStatus != nil && healthStatus.Status == health.HealthStatusHealthy {
					ticker.Stop()
					return nil
				}
			}
		}
	}
}

func (deployment *Deployment) watchDeploy() error {
	err := deployment.Deploy()
	if err != nil {
		return customerror.WithOperation(err, "deployment.deploy.Deploy")
	}

	err = deployment.NewWatcher()
	if err != nil {
		return customerror.WithOperation(err, "deployment.deploy.NewWatcher")
	}

	return nil
}

func (deployment *Deployment) Deploy() error {
	manifest := deployment.getUnstructuredManifest()

	_, err := deployment.kubectl.ApplyResource(
		context.TODO(),
		deployment.config,
		manifest,
		deployment.namespace,
		util.DryRunNone,
		true,
		false,
	)

	if err != nil {
		return customerror.New(
			"Deploy failed",
			err.Error(),
			map[string]string{
				"resourceName":     manifest.GetName(),
				"groupVersionKind": manifest.GroupVersionKind().String(),
			},
			"deployment.deploy.ApplyResource",
		)
	}

	return deployment.NewWatcher()
}

func isResourController(resource *unstructured.Unstructured) bool {
	if resource == nil {
		return false
	}

	_, isRsourceController, _ := unstructured.NestedInt64(resource.Object, "spec", "replicas")
	return isRsourceController
}

func (deployment *Deployment) Undeploy() error {
	manifest := deployment.getUnstructuredManifest()
	gvk := manifest.GroupVersionKind()

	resourceInCluster, err := deployment.kubectl.GetResource(context.TODO(), deployment.config, gvk, manifest.GetName(), deployment.namespace)
	if err != nil && k8sErrors.IsNotFound(err) {
		return nil
	}

	if err != nil {
		return customerror.New(
			"Undeploy failed",
			err.Error(),
			map[string]string{
				"resourceName":     manifest.GetName(),
				"groupVersionKind": manifest.GroupVersionKind().String(),
			},
			"deployment.undeploy.GetResource",
		)
	}

	if !isResourController(resourceInCluster) {
		return nil
	}

	err = deployment.kubectl.DeleteResource(context.TODO(), deployment.config, gvk, manifest.GetName(), deployment.namespace, true)
	if err != nil && k8sErrors.IsNotFound(err) {
		return nil
	}

	if err != nil {
		return customerror.New(
			"Undeploy failed",
			err.Error(),
			map[string]string{
				"resourceName":     manifest.GetName(),
				"groupVersionKind": manifest.GroupVersionKind().String(),
			},
			"deployment.undeploy.DeleteResource",
		)
	}

	return nil
}

func (deployment *Deployment) getUnstructuredManifest() *unstructured.Unstructured {
	return &unstructured.Unstructured{
		Object: deployment.manifest,
	}
}

File: ui/.stryker-tmp/sandbox3301781/src/core/providers/deployment.ts
/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
// @ts-nocheck

function stryNS_9fa48() {
  var g = new Function("return this")();
  var ns = g.__stryker__ || (g.__stryker__ = {});

  if (ns.activeMutant === undefined && g.process && g.process.env && g.process.env.__STRYKER_ACTIVE_MUTANT__) {
    ns.activeMutant = g.process.env.__STRYKER_ACTIVE_MUTANT__;
  }

  function retrieveNS() {
    return ns;
  }

  stryNS_9fa48 = retrieveNS;
  return retrieveNS();
}

stryNS_9fa48();

function stryCov_9fa48() {
  var ns = stryNS_9fa48();
  var cov = ns.mutantCoverage || (ns.mutantCoverage = {
    static: {},
    perTest: {}
  });

  function cover() {
    var c = cov.static;

    if (ns.currentTestId) {
      c = cov.perTest[ns.currentTestId] = cov.perTest[ns.currentTestId] || {};
    }

    var a = arguments;

    for (var i = 0; i < a.length; i++) {
      c[a[i]] = (c[a[i]] || 0) + 1;
    }
  }

  stryCov_9fa48 = cover;
  cover.apply(null, arguments);
}

function stryMutAct_9fa48(id) {
  var ns = stryNS_9fa48();

  function isActive(id) {
    if (ns.activeMutant === id) {
      if (ns.hitCount !== void 0 && ++ns.hitCount > ns.hitLimit) {
        throw new Error('Stryker: Hit count limit reached (' + ns.hitCount + ')');
      }

      return true;
    }

    return false;
  }

  stryMutAct_9fa48 = isActive;
  return isActive(id);
}

import { CreateDeployment } from 'modules/Circles/Release/interfaces/Deployment';
import { postRequest, baseRequest } from './base';
const v2Endpoint = stryMutAct_9fa48("1670") ? "" : (stryCov_9fa48("1670"), '/moove/v2/deployments');
export const createDeployment = stryMutAct_9fa48("1671") ? () => undefined : (stryCov_9fa48("1671"), (() => {
  const createDeployment = (data: CreateDeployment) => postRequest(v2Endpoint, data);

  return createDeployment;
})());
export const undeploy = stryMutAct_9fa48("1672") ? () => undefined : (stryCov_9fa48("1672"), (() => {
  const undeploy = (deploymentId: string) => postRequest(stryMutAct_9fa48("1673") ? `` : (stryCov_9fa48("1673"), `${v2Endpoint}/${deploymentId}/undeploy`));

  return undeploy;
})());
export const getDeployHistoryByCircleId = (params: URLSearchParams, circleId: string) => {
  if (stryMutAct_9fa48("1674")) {
    {}
  } else {
    stryCov_9fa48("1674");
    params.append(stryMutAct_9fa48("1675") ? "" : (stryCov_9fa48("1675"), 'size'), stryMutAct_9fa48("1676") ? "" : (stryCov_9fa48("1676"), '10'));
    return baseRequest(stryMutAct_9fa48("1677") ? `` : (stryCov_9fa48("1677"), `${v2Endpoint}/circle/${circleId}/history?${params}`));
  }
};
export const findDeployLogsByDeploymentId = stryMutAct_9fa48("1678") ? () => undefined : (stryCov_9fa48("1678"), (() => {
  const findDeployLogsByDeploymentId = (deploymentId: string) => baseRequest(stryMutAct_9fa48("1679") ? `` : (stryCov_9fa48("1679"), `${v2Endpoint}/${deploymentId}/logs`));

  return findDeployLogsByDeploymentId;
})());